Design/Rule0018NoEventsInInternalCodeunits.cs

```diff
@@ -0,0 +1,84 @@
+﻿using Microsoft.Dynamics.Nav.Analyzers.Common.AppSourceCopConfiguration;
+using Microsoft.Dynamics.Nav.CodeAnalysis;
+using Microsoft.Dynamics.Nav.CodeAnalysis.Diagnostics;
+using Microsoft.Dynamics.Nav.CodeAnalysis.InternalSyntax;
+using Microsoft.Dynamics.Nav.CodeAnalysis.Symbols;
+using System;
+using System.Collections.Immutable;
+
+namespace BusinessCentral.LinterCop.Design {
+    [DiagnosticAnalyzer]
+    class Rule0018NoEventsInInternalCodeunits : DiagnosticAnalyzer {
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create<DiagnosticDescriptor>(DiagnosticDescriptors.Rule0018NoEventsInInternalCodeunitsAnalyzerDescriptor);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSymbolAction(new Action<SymbolAnalysisContext>(CheckInternalCodeunitForEvents), SymbolKind.Codeunit);
+        }
+
+        private void CheckInternalCodeunitForEvents(SymbolAnalysisContext symbolAnalysisContext)
+        {
+            ICodeunitTypeSymbol codeunitSymbol = symbolAnalysisContext.Symbol as ICodeunitTypeSymbol;
+            if (codeunitSymbol == null)
+            {
+                return;
+            }
+            if (codeunitSymbol.IsObsoleteRemoved || codeunitSymbol.IsObsoletePending)
+            {
+                return;
+            }
+            if (codeunitSymbol.DeclaredAccessibility != Accessibility.Internal)
+            {
+                return;
+            }
+
+            ImmutableArray<ISymbol>.Enumerator enumerator = codeunitSymbol.GetMembers().GetEnumerator();
+            while (enumerator.MoveNext())
+            {
+                ISymbol current = enumerator.Current;
+                SymbolKind kind = current.Kind;
+
+                if (kind != SymbolKind.Method)
+                {
+                    continue;
+                }
+
+                IMethodSymbol methodSymbol = (IMethodSymbol)current;
+                if (!methodSymbol.IsEvent)
+                {
+                    continue;
+                }
+
+                IAttributeSymbol attributeSymbol;
+                if (!TryGetEventAttribute(methodSymbol, out attributeSymbol))
+                {
+                    continue;
+                }
+                if (attributeSymbol.AttributeKind == AttributeKind.InternalEvent)
+                {
+                    continue;
+                }
+
+                symbolAnalysisContext.ReportDiagnostic(Diagnostic.Create(DiagnosticDescriptors.Rule0018NoEventsInInternalCodeunitsAnalyzerDescriptor, methodSymbol.GetLocation(), new Object[] { methodSymbol.Name, codeunitSymbol.Name }));
+            }
+        }
+
+        private bool TryGetEventAttribute(IMethodSymbol method, out IAttributeSymbol attribute)
+        {
+            ImmutableArray<IAttributeSymbol>.Enumerator enumerator = method.Attributes.GetEnumerator();
```

Comment:
I'm a bit concerned about performance of LinterCop. Do you know if iterating through a an enumerator is the fastest way of doing this? As the code is now, there is iterations inside iterations, which is usually something we try to avoid in BC - and probably in C# as well. 🙂

(I haven't checked the implementations of other rules, it was just a reflection while looking through your PR out of curiosity)

Reply:
And maybe this is not the place for this kind of question, since it is more a generic performance question. 🙂

But I took a quick look at a few other rules. I saw a couple of iterating enumerators in existing code as well, but I also noted that the `enumerable.where(x => condition on x)` was used now and then. The later looks more slim to me, but I do not know if it performs better... Do you?

Reply:
I also prefer the later and I will need to refactor the older rules at some point :)